text/3714-macro-fragment-fields.md

```diff
@@ -0,0 +1,182 @@
+of the matched type with `${t.name}`:
+
+```rust
+macro_rules! get_name {
+    ($t:adt) => { println!("{}", stringify!(${t.name})); }
+}
+
+fn main() {
+    let n1 = get_name!(struct S { field: u32 });
+    let n2 = get_name!(enum E { V1, V2 = 42, V3(u8) });
+    let n3 = get_name!(union U { u: u32, f: f32 });
+    println!("{n3}{n1}{n2}"); // prints "USE"
+}
+```
+
+An attempt to access a field that doesn't exist will produce a compilation
+error on the macro definition, whether or not the specific macro rule gets
+invoked.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+Fragment fields may be used in a macro transcriber anywhere the corresponding
+fragment name could be used.
+
+Fragment fields typically follow the same rules for repetition handling as the
+corresponding fragment (e.g. being used at the same level/kind of repetition).
+However, fragment fields that contain multiple items require one additional
+level of repetition; see the `param` field of `:fn`, below.
+
+This RFC introduces the following new fragment specifiers, with specified fields:
+
+- `:fn`: A function definition (including body).
+  - `name`: The name of the function, as an `ident`.
+  - `param`: The parameters of the function, presented as though captured by a
+    level of `*` repetition. For instance, you can write `$(${f.param}),*` to
+    get a comma-separated list of parameters, or `$(other_macro!(${f.param}))*`
+    to pass each parameter to another macro.
+  - `return_type`: The return type of the function, as a `ty`. If the function
+    has no explicitly specified return type, this will be `()`, with a span of
+    the closing parenthesis for the function arguments.
+  - `body`: The body of the function, as a block (including the
+    surrounding braces).
+  - `vis`: The visibility of the function, as a `vis` (may be empty).
+- `:adt`: An ADT (struct, union, or enum).
+  - `name`: The name of the ADT, as an `ident`.
+
+The tokens within fields have the spans of the corresponding tokens from the
+source. If a token has no corresponding source (e.g. the `()` in `return_type`
+for a `fn` with no explicitly specified return type), the field definition
+defines an appropriate span.
+
+Using a field of a fragment counts as a use of the fragment, for the purposes
+of ensuring every fragment gets used at least once at the appropriate level of
+repetition.
+
+This extends the grammar of macro metavariable expressions to allow using a dot
+and identifier to access a field.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+This adds complexity to the macro system, in order to simplify macros in the
+ecosystem.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+Rather than using field syntax, we could use function-like syntax in the style
+of RFC 3086's macro metavariable expressions. However, field syntax seems like
+a more natural fit for this concept.
```

Comment:
Something this made me realize is we could all spell the fields as built-in macros, like `adt_name!`. You'd end up with something like:

```rust
macro_rules! get_name {
    ($t:adt) => { println!("{}", stringify!(adt_name!($t))); }
}

fn main() {
    let n1 = get_name!(struct S { field: u32 });
    let n2 = get_name!(enum E { V1, V2 = 42, V3(u8) });
    let n3 = get_name!(union U { u: u32, f: f32 });
    println!("{n3}{n1}{n2}"); // prints "USE"
}
```

The field syntax is probably more ergonomic, but the built-in macro approach would save us from introducing a new concept (fragment fields).

It also seems like we could prototype most of this with a big bag of `macro_rules!` macros in a crate. 

The new fragments (`adt`, `fn`, etc.) seem pretty useful though, and harder to do as a library.

Reply:
@eholk result of macro calls can't be feed into another macro (`stringify` will literally see `["adt_name", "!", "(", $t, ")"]`), but result of metavariable expressions can.

```rust
#![feature(macro_metavar_expr)]
macro_rules! expect_literal {
    ($a:literal) => {}
}

macro_rules! count {
    ($($x:tt)*) => { ${count($x)} }
}

macro_rules! good {
    ($($x:tt)*) => { expect_literal!(${count($x)}) }
}

macro_rules! bad {
    ($($x:tt)*) => { expect_literal!(count!($($x)*)) }
//                                   ^^^^^ no rules expected this token in macro call
}

fn main() {
    good!(1 2 3);
    bad!(1 2 3);
}
```